File: zig/src/queries/t3.zig
// queries: 7a-c, 8a-d, 9a-d, 10a-c (queries.jl lines 591-753)
const std = @import("std");
const Io = std.Io;
const Data = @import("../data.zig").Data;
const h = @import("../helpers.zig");
const rx = @import("../regex.zig");
const sets = @import("../sets.zig");
const Entry = @import("all.zig").Entry;

pub const ENTRIES: []const Entry = &.{
    .{ .name = "7a",  .oracle = "Antonioni, Michelangelo || Dressed to Kill",                                 .run = q7a },
    .{ .name = "7b",  .oracle = "De Palma, Brian || Dressed to Kill",                                         .run = q7b },
    .{ .name = "7c",  .oracle = "50 Cent || \"Boo\" Arnold was born Earl Arnold in Hattiesburg, Mississippi in 1966. His father gave him the nickname 'Boo' early in life and it stuck through grade school, high school, and college. He is still known as \"Boo\" to family and friends.  Raised in central Texas, Arnold played baseball at Texas Tech University where he graduated with a BA in Advertising and Marketing. While at Texas Tech he was also a member of the Texas Epsilon chapter of Phi Delta Theta fraternity. After college he worked with Young Life, an outreach to high school students, in San Antonio, Texas.  While with Young Life Arnold began taking extension courses through Fuller Theological Seminary and ultimately went full-time to Gordon-Conwell Theological Seminary in Boston, Massachusetts. At Gordon-Conwell he completed a Master's Degree in Divinity studying Theology, Philosophy, Church History, Biblical Languages (Hebrew & Greek), and Exegetical Methods. Following seminary he was involved with reconciliation efforts in the former Yugoslavia shortly after the war ended there in1995.  Arnold started acting in his early thirties in Texas. After an encouraging visit to Los Angeles where he spent time with childhood friend George Eads (of CSI Las Vegas) he decided to move to Los Angeles in 2001 to pursue acting full-time. While in Los Angeles he has studied acting with Judith Weston at Judith Weston Studio for Actors and Directors.  Arnold's acting career has been one of steady development, booking co-star and guest-star roles in nighttime television. He guest-starred opposite of Jane Seymour on the night time television drama Justice. He played the lead, Michael Hollister, in the film The Seer, written and directed by Patrick Masset (Friday Night Lights).  He was nominated Best Actor in the168 Film Festival for the role of Phil Stevens in the short-film Useless. In Useless he played a US Marshal who must choose between mercy and justice as he confronts the man who murdered his father. Arnold's performance in Useless confirmed his ability to carry lead roles, and he continues to work toward solidifying himself as a male lead in film and television.  Arnold married fellow Texan Stacy Rudd of San Antonio in 2003 and they are now raising their three children in the Los Angeles area.", .run = q7c },
    .{ .name = "8a",  .oracle = "Chambers, Linda || .hack//Quantum",                                          .run = q8a },
    .{ .name = "8b",  .oracle = "Chambers, Linda || Dragon Ball Z: Shin Budokai",                             .run = q8b },
    .{ .name = "8c",  .oracle = "\"A.J.\" || #1 Cheerleader Camp",                                            .run = q8c },
    .{ .name = "8d",  .oracle = "\"Jenny from the Block\" || #1 Cheerleader Camp",                            .run = q8d },
    .{ .name = "9a",  .oracle = "AJ || Airport Announcer || Blue Harvest",                                    .run = q9a },
    .{ .name = "9b",  .oracle = "AJ || Airport Announcer || Bassett, Angela || Blue Harvest",                 .run = q9b },
    .{ .name = "9c",  .oracle = "'Annette' || 2nd Balladeer || Alborg, Ana Esther || (1975-01-20)",           .run = q9c },
    .{ .name = "9d",  .oracle = "!!!, Toy || Aaron, Caroline || \"Cockamamie's\" Salesgirl || $15,000.00 Error", .run = q9d },
    .{ .name = "10a", .oracle = "Actor || 12 Rounds",                                                         .run = q10a },
    .{ .name = "10b", .oracle = "(empty)",                                                                    .run = q10b },
    .{ .name = "10c", .oracle = "Himself || Evil Eyes: Behind the Scenes",                                    .run = q10c },
};

fn q7a(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_production_year.ge(1980).k()
            .@"and"(d.movie_production_year.le(1995).k())
            .@"and"(d.movie_linked_by.o(
                d.movielink_type.o(d.linktype_link).eq("features")).k())
            .o(
                d.movie_cast.o(
                    d.cast_person.in_s(
                        d.person_aka.o(d.akaname_name).rx(rx.a_lc).k()
                            .@"and"(d.person_name_pcode.ge("A").k())
                            .@"and"(d.person_name_pcode.le("F").k())
                            .@"and"(
                                d.person_gender.eq("m").k()
                                    .@"or"(
                                        d.person_gender.eq("f").k()
                                            .@"and"(d.person_name.rx(rx.pre_B).k())
                                    )
                            )
                            .@"and"(d.person_info.in_s(
                                d.personinfo_type.o(d.infotype_info).eq("mini biography").k()
                                    .@"and"(d.personinfo_note.eq("Volker Boehm").k())
                            ).k())
                    ).k().o(d.cast_person.o(d.person_name))
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc2{};
    q.drive(h.Sink(h.Acc2){ .acc = &acc });
    try h.fmt2(w, acc.m0, acc.m1);
}

fn q7b(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_production_year.ge(1980).k()
            .@"and"(d.movie_production_year.le(1984).k())
            .@"and"(d.movie_linked_by.o(
                d.movielink_type.o(d.linktype_link).eq("features")).k())
            .o(
                d.movie_cast.o(
                    d.cast_person.in_s(
                        d.person_aka.o(d.akaname_name).rx(rx.a_lc).k()
                            .@"and"(d.person_name_pcode.rx(rx.pre_D).k())
                            .@"and"(d.person_gender.eq("m").k())
                            .@"and"(d.person_info.in_s(
                                d.personinfo_type.o(d.infotype_info).eq("mini biography").k()
                                    .@"and"(d.personinfo_note.eq("Volker Boehm").k())
                            ).k())
                    ).k().o(d.cast_person.o(d.person_name))
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc2{};
    q.drive(h.Sink(h.Acc2){ .acc = &acc });
    try h.fmt2(w, acc.m0, acc.m1);
}

fn q7c(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_production_year.ge(1980).k()
            .@"and"(d.movie_production_year.le(2010).k())
            .@"and"(d.movie_linked_by.o(
                d.movielink_type.o(d.linktype_link).in_v(
                    &[_][]const u8{ "references", "referenced in", "features", "featured in" })).k())
            .o(
                d.movie_cast.o(
                    d.cast_person.in_s(
                        d.person_aka.o(d.akaname_name).rx(rx.a_or_pre_A).k()
                            .@"and"(d.person_name_pcode.ge("A").k())
                            .@"and"(d.person_name_pcode.le("F").k())
                            .@"and"(
                                d.person_gender.eq("m").k()
                                    .@"or"(
                                        d.person_gender.eq("f").k()
                                            .@"and"(d.person_name.rx(rx.pre_A).k())
                                    )
                            )
                            .@"and"(d.person_info.in_s(
                                d.personinfo_type.o(d.infotype_info).eq("mini biography").k()
                                    .@"and"(d.personinfo_note.k())
                            ).k())
                    ).k().o(
                        d.cast_person.o(d.person_name)
                            .x(d.cast_person.o(
                                d.person_info.o(
                                    d.personinfo_type.o(d.infotype_info).eq("mini biography").k()
                                        .@"and"(d.personinfo_note.k())
                                        .o(d.personinfo_info))))
                    )
                )
            )
    );
    var acc = h.Acc2{};
    q.drive(h.Sink(h.Acc2){ .acc = &acc });
    try h.fmt2(w, acc.m0, acc.m1);
}

fn q8a(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_company.in_s(
            d.company_country.eq("[jp]").k()
                .@"and"(d.company_note.rx(rx.paren_japan).k())
                .@"and"(d.company_note.nrx(rx.paren_USA).k())
        ).k()
            .o(
                d.movie_cast.o(
                    d.cast_note.eq("(voice: English version)").k()
                        .@"and"(d.cast_role.o(d.roletype_role).eq("actress").k())
                        .@"and"(d.cast_person.in_s(
                            d.person_name.rx(rx.yo).k()
                                .@"and"(d.person_name.nrx(rx.yu).k())
                        ).k())
                        .o(d.cast_person.o(d.person_aka.o(d.akaname_name)))
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc2{};
    q.drive(h.Sink(h.Acc2){ .acc = &acc });
    try h.fmt2(w, acc.m0, acc.m1);
}

fn q8b(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_company.in_s(
            d.company_country.eq("[jp]").k()
                .@"and"(d.company_note.rx(rx.paren_japan).k())
                .@"and"(d.company_note.nrx(rx.paren_USA).k())
                .@"and"(
                    d.company_note.rx(rx.paren_2006).k()
                        .@"or"(d.company_note.rx(rx.paren_2007).k())
                )
        ).k()
            .@"and"(d.movie_production_year.ge(2006).k())
            .@"and"(d.movie_production_year.le(2007).k())
            .@"and"(
                d.movie_title.rx(rx.pre_one_piece).k()
                    .@"or"(d.movie_title.rx(rx.pre_dragon_ball_z).k())
            )
            .o(
                d.movie_cast.o(
                    d.cast_note.eq("(voice: English version)").k()
                        .@"and"(d.cast_role.o(d.roletype_role).eq("actress").k())
                        .@"and"(d.cast_person.in_s(
                            d.person_name.rx(rx.yo).k()
                                .@"and"(d.person_name.nrx(rx.yu).k())
                        ).k())
                        .o(d.cast_person.o(d.person_aka.o(d.akaname_name)))
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc2{};
    q.drive(h.Sink(h.Acc2){ .acc = &acc });
    try h.fmt2(w, acc.m0, acc.m1);
}

fn q8c(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_company.o(d.company_country.eq("[us]")).k()
            .o(
                d.movie_cast.o(
                    d.cast_role.o(d.roletype_role).eq("writer").k()
                        .o(d.cast_person.o(d.person_aka.o(d.akaname_name)))
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc2{};
    q.drive(h.Sink(h.Acc2){ .acc = &acc });
    try h.fmt2(w, acc.m0, acc.m1);
}

fn q8d(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_company.o(d.company_country.eq("[us]")).k()
            .o(
                d.movie_cast.o(
                    d.cast_role.o(d.roletype_role).eq("costume designer").k()
                        .o(d.cast_person.o(d.person_aka.o(d.akaname_name)))
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc2{};
    q.drive(h.Sink(h.Acc2){ .acc = &acc });
    try h.fmt2(w, acc.m0, acc.m1);
}

fn q9a(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_company.in_s(
            d.company_country.eq("[us]").k()
                .@"and"(
                    d.company_note.rx(rx.paren_USA).k()
                        .@"or"(d.company_note.rx(rx.paren_worldwide).k())
                )
        ).k()
            .@"and"(d.movie_production_year.ge(2005).k())
            .@"and"(d.movie_production_year.le(2015).k())
            .o(
                d.movie_cast.o(
                    d.cast_note.in_v(sets.voice4).k()
                        .@"and"(d.cast_role.o(d.roletype_role).eq("actress").k())
                        .@"and"(d.cast_person.in_s(
                            d.person_gender.eq("f").k()
                                .@"and"(d.person_name.rx(rx.ang).k())
                        ).k())
                        .o(
                            d.cast_person.o(d.person_aka.o(d.akaname_name))
                                .x(d.cast_character.o(d.character_name))
                        )
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc3{};
    q.drive(h.Sink(h.Acc3){ .acc = &acc });
    try h.fmt3(w, acc.m0, acc.m1, acc.m2);
}

fn q9b(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_company.in_s(
            d.company_country.eq("[us]").k()
                .@"and"(d.company_note.rx(rx.paren_200_dot).k())
                .@"and"(
                    d.company_note.rx(rx.paren_USA).k()
                        .@"or"(d.company_note.rx(rx.paren_worldwide).k())
                )
        ).k()
            .@"and"(d.movie_production_year.ge(2007).k())
            .@"and"(d.movie_production_year.le(2010).k())
            .o(
                d.movie_cast.o(
                    d.cast_note.eq("(voice)").k()
                        .@"and"(d.cast_role.o(d.roletype_role).eq("actress").k())
                        .@"and"(d.cast_person.in_s(
                            d.person_gender.eq("f").k()
                                .@"and"(d.person_name.rx(rx.angel).k())
                        ).k())
                        .o(
                            d.cast_person.o(d.person_aka.o(d.akaname_name))
                                .x(d.cast_character.o(d.character_name))
                                .x(d.cast_person.o(d.person_name))
                        )
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc4{};
    q.drive(h.Sink(h.Acc4){ .acc = &acc });
    try h.fmt4(w, acc.m0, acc.m1, acc.m2, acc.m3);
}

fn q9c(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_company.o(d.company_country.eq("[us]")).k()
            .o(
                d.movie_cast.o(
                    d.cast_note.in_v(sets.voice4).k()
                        .@"and"(d.cast_role.o(d.roletype_role).eq("actress").k())
                        .@"and"(d.cast_person.in_s(
                            d.person_gender.eq("f").k()
                                .@"and"(d.person_name.rx(rx.an_uc).k())
                        ).k())
                        .o(
                            d.cast_person.o(d.person_aka.o(d.akaname_name))
                                .x(d.cast_character.o(d.character_name))
                                .x(d.cast_person.o(d.person_name))
                        )
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc4{};
    q.drive(h.Sink(h.Acc4){ .acc = &acc });
    try h.fmt4(w, acc.m0, acc.m1, acc.m2, acc.m3);
}

fn q9d(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_company.o(d.company_country.eq("[us]")).k()
            .o(
                d.movie_cast.o(
                    d.cast_note.in_v(sets.voice4).k()
                        .@"and"(d.cast_role.o(d.roletype_role).eq("actress").k())
                        .@"and"(d.cast_person.o(d.person_gender.eq("f")).k())
                        .o(
                            d.cast_person.o(d.person_aka.o(d.akaname_name))
                                .x(d.cast_person.o(d.person_name))
                                .x(d.cast_character.o(d.character_name))
                        )
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc4{};
    q.drive(h.Sink(h.Acc4){ .acc = &acc });
    try h.fmt4(w, acc.m0, acc.m1, acc.m2, acc.m3);
}

fn q10a(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_company.o(d.company_country.eq("[ru]")).k()
            .@"and"(d.movie_production_year.gt(2005).k())
            .o(
                d.movie_cast.o(
                    d.cast_note.rx(rx.paren_voice).k()
                        .@"and"(d.cast_note.rx(rx.paren_uncredited).k())
                        .@"and"(d.cast_role.o(d.roletype_role).eq("actor").k())
                        .o(d.cast_character.o(d.character_name))
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc2{};
    q.drive(h.Sink(h.Acc2){ .acc = &acc });
    try h.fmt2(w, acc.m0, acc.m1);
}

fn q10b(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_company.o(d.company_country.eq("[ru]")).k()
            .@"and"(d.movie_production_year.gt(2010).k())
            .o(
                d.movie_cast.o(
                    d.cast_note.rx(rx.paren_producer).k()
                        .@"and"(d.cast_role.o(d.roletype_role).eq("actor").k())
                        .o(d.cast_character.o(d.character_name))
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc2{};
    q.drive(h.Sink(h.Acc2){ .acc = &acc });
    try h.fmt2(w, acc.m0, acc.m1);
}

fn q10c(d: *const Data, w: *Io.Writer) anyerror!void {
    const q = d.movie.o(
        d.movie_company.o(d.company_country.eq("[us]")).k()
            .@"and"(d.movie_production_year.gt(1990).k())
            .o(
                d.movie_cast.o(
                    d.cast_note.rx(rx.paren_producer).k()
                        .o(d.cast_character.o(d.character_name))
                ).x(d.movie_title)
            )
    );
    var acc = h.Acc2{};
    q.drive(h.Sink(h.Acc2){ .acc = &acc });
    try h.fmt2(w, acc.m0, acc.m1);
}

File: rust/src/queries/t3.rs
// queries: 7a-c, 8a-d, 9a-d, 10a-c (queries.jl lines 591-753)

use crate::engine::*;
use crate::job_schema::*;
use crate::queries::helpers::min_row;
use crate::queries::sets::voice4;

pub const ENTRIES: &[super::Entry] = &[
    ("7a",  "Antonioni, Michelangelo || Dressed to Kill",                                 q7a),
    ("7b",  "De Palma, Brian || Dressed to Kill",                                         q7b),
    ("7c",  "50 Cent || \"Boo\" Arnold was born Earl Arnold in Hattiesburg, Mississippi in 1966. His father gave him the nickname 'Boo' early in life and it stuck through grade school, high school, and college. He is still known as \"Boo\" to family and friends.  Raised in central Texas, Arnold played baseball at Texas Tech University where he graduated with a BA in Advertising and Marketing. While at Texas Tech he was also a member of the Texas Epsilon chapter of Phi Delta Theta fraternity. After college he worked with Young Life, an outreach to high school students, in San Antonio, Texas.  While with Young Life Arnold began taking extension courses through Fuller Theological Seminary and ultimately went full-time to Gordon-Conwell Theological Seminary in Boston, Massachusetts. At Gordon-Conwell he completed a Master's Degree in Divinity studying Theology, Philosophy, Church History, Biblical Languages (Hebrew & Greek), and Exegetical Methods. Following seminary he was involved with reconciliation efforts in the former Yugoslavia shortly after the war ended there in1995.  Arnold started acting in his early thirties in Texas. After an encouraging visit to Los Angeles where he spent time with childhood friend George Eads (of CSI Las Vegas) he decided to move to Los Angeles in 2001 to pursue acting full-time. While in Los Angeles he has studied acting with Judith Weston at Judith Weston Studio for Actors and Directors.  Arnold's acting career has been one of steady development, booking co-star and guest-star roles in nighttime television. He guest-starred opposite of Jane Seymour on the night time television drama Justice. He played the lead, Michael Hollister, in the film The Seer, written and directed by Patrick Masset (Friday Night Lights).  He was nominated Best Actor in the168 Film Festival for the role of Phil Stevens in the short-film Useless. In Useless he played a US Marshal who must choose between mercy and justice as he confronts the man who murdered his father. Arnold's performance in Useless confirmed his ability to carry lead roles, and he continues to work toward solidifying himself as a male lead in film and television.  Arnold married fellow Texan Stacy Rudd of San Antonio in 2003 and they are now raising their three children in the Los Angeles area.", q7c),
    ("8a",  "Chambers, Linda || .hack//Quantum",                                          q8a),
    ("8b",  "Chambers, Linda || Dragon Ball Z: Shin Budokai",                             q8b),
    ("8c",  "\"A.J.\" || #1 Cheerleader Camp",                                            q8c),
    ("8d",  "\"Jenny from the Block\" || #1 Cheerleader Camp",                            q8d),
    ("9a",  "AJ || Airport Announcer || Blue Harvest",                                    q9a),
    ("9b",  "AJ || Airport Announcer || Bassett, Angela || Blue Harvest",                 q9b),
    ("9c",  "'Annette' || 2nd Balladeer || Alborg, Ana Esther || (1975-01-20)",           q9c),
    ("9d",  "!!!, Toy || Aaron, Caroline || \"Cockamamie's\" Salesgirl || $15,000.00 Error", q9d),
    ("10a", "Actor || 12 Rounds",                                                         q10a),
    ("10b", "(empty)",                                                                    q10b),
    ("10c", "Himself || Evil Eyes: Behind the Scenes",                                    q10c),
];

fn q7a() -> String {
    min_row(movie()
        .when(production_year().ge(1980)
              .and(production_year().le(1995))
              .and(linked_by().ty().text().eq("features")))
        .get(cast().get(person()
                        .when(alias().text().rx(r"a")
                              .and(name_pcode_cf().ge("A"))
                              .and(name_pcode_cf().le("F"))
                              .and(gender().eq("m")
                                   .or(gender().eq("f").and(Person::name().rx(r"^B"))))
                              .and(bio().get(PersonInfo::ty().text().eq("mini biography")
                                             .and(PersonInfo::note().eq("Volker Boehm")))))
                        .name())
             .and(title())))
}

fn q7b() -> String {
    min_row(movie()
        .when(production_year().ge(1980)
              .and(production_year().le(1984))
              .and(linked_by().ty().text().eq("features")))
        .get(cast().get(person()
                        .when(alias().text().rx(r"a")
                              .and(name_pcode_cf().rx(r"^D"))
                              .and(gender().eq("m"))
                              .and(bio().get(PersonInfo::ty().text().eq("mini biography")
                                             .and(PersonInfo::note().eq("Volker Boehm")))))
                        .name())
             .and(title())))
}

// Conjunct tree (∧ = Prod) — consumed via `member` only, so the value
// type stays opaque (`impl Query<D = Id<PersonInfo>> + Probe`).
fn bio_filter_7c() -> impl Query<D = Id<PersonInfo>> + Probe {
    PersonInfo::ty().text().eq("mini biography")
        .and(PersonInfo::note())
}

fn q7c() -> String {
    min_row(movie()
        .when(production_year().ge(1980)
              .and(production_year().le(2010))
              .and(linked_by().ty().text().is_in(
                  ["references", "referenced in", "features", "featured in"])))
        .get(cast().get(person()
                        .when(alias().text().rx(r"a|^A")
                              .and(name_pcode_cf().ge("A"))
                              .and(name_pcode_cf().le("F"))
                              .and(gender().eq("m")
                                   .or(gender().eq("f").and(Person::name().rx(r"^A"))))
                              .and(bio().when(bio_filter_7c())))
                        .get(Person::name()
                             .and(bio().when(bio_filter_7c()).info())))))
}

fn q8a() -> String {
    min_row(movie()
        .when(company().get(country().eq("[jp]")
                            .and(Company::note().rx(r"\(Japan\)"))
                            .and(Company::note().nrx(r"\(USA\)"))))
        .get(cast()
             .when(Cast::note().eq("(voice: English version)")
                   .and(role().text().eq("actress"))
                   .and(person().when(Person::name().rx(r"Yo")
                                      .and(Person::name().nrx(r"Yu")))))
             .person().alias().text()
             .and(title())))
}

fn q8b() -> String {
    min_row(movie()
        .when(company().get(country().eq("[jp]")
                            .and(Company::note().rx(r"\(Japan\)"))
                            .and(Company::note().nrx(r"\(USA\)"))
                            .and(Company::note().rx(r"\(2006\)")
                                 .or(Company::note().rx(r"\(2007\)"))))
              .and(production_year().ge(2006))
              .and(production_year().le(2007))
              .and(title().rx(r"^One Piece").or(title().rx(r"^Dragon Ball Z"))))
        .get(cast()
             .when(Cast::note().eq("(voice: English version)")
                   .and(role().text().eq("actress"))
                   .and(person().when(Person::name().rx(r"Yo")
                                      .and(Person::name().nrx(r"Yu")))))
             .person().alias().text()
             .and(title())))
}

// q8c/q8d differ only in the cast role.
fn q8cd(role_: &'static str) -> String {
    min_row(movie()
        .when(company().country().eq("[us]"))
        .get(cast().when(role().text().eq(role_))
             .person().alias().text()
             .and(title())))
}

fn q8c() -> String { q8cd("writer") }
fn q8d() -> String { q8cd("costume designer") }

fn q9a() -> String {
    min_row(movie()
        .when(company().get(country().eq("[us]")
                            .and(Company::note().rx(r"\(USA\)")
                                 .or(Company::note().rx(r"\(worldwide\)"))))
              .and(production_year().ge(2005))
              .and(production_year().le(2015)))
        .get(cast()
             .when(Cast::note().is_in(voice4())
                   .and(role().text().eq("actress"))
                   .and(person().when(gender().eq("f")
                                      .and(Person::name().rx(r"Ang")))))
             .get(person().alias().text()
                  .and(character().text()))
             .and(title())))
}

fn q9b() -> String {
    min_row(movie()
        .when(company().get(country().eq("[us]")
                            .and(Company::note().rx(r"\(200.*\)"))
                            .and(Company::note().rx(r"\(USA\)")
                                 .or(Company::note().rx(r"\(worldwide\)"))))
              .and(production_year().ge(2007))
              .and(production_year().le(2010)))
        .get(cast()
             .when(Cast::note().eq("(voice)")
                   .and(role().text().eq("actress"))
                   .and(person().when(gender().eq("f")
                                      .and(Person::name().rx(r"Angel")))))
             .get(person().alias().text()
                  .and(character().text())
                  .and(person().name()))
             .and(title())))
}

fn q9c() -> String {
    min_row(movie()
        .when(company().country().eq("[us]"))
        .get(cast()
             .when(Cast::note().is_in(voice4())
                   .and(role().text().eq("actress"))
                   .and(person().when(gender().eq("f")
                                      .and(Person::name().rx(r"An")))))
             .get(person().alias().text()
                  .and(character().text())
                  .and(person().name()))
             .and(title())))
}

fn q9d() -> String {
    min_row(movie()
        .when(company().country().eq("[us]"))
        .get(cast()
             .when(Cast::note().is_in(voice4())
                   .and(role().text().eq("actress"))
                   .and(person().when(gender().eq("f"))))
             .get(person().alias().text()
                  .and(person().name())
                  .and(character().text()))
             .and(title())))
}

fn q10a() -> String {
    min_row(movie()
        .when(company().country().eq("[ru]")
              .and(production_year().gt(2005)))
        .get(cast()
             .when(Cast::note().rx(r"\(voice\)")
                   .and(Cast::note().rx(r"\(uncredited\)"))
                   .and(role().text().eq("actor")))
             .character().text()
             .and(title())))
}

fn q10b() -> String {
    min_row(movie()
        .when(company().country().eq("[ru]")
              .and(production_year().gt(2010)))
        .get(cast()
             .when(Cast::note().rx(r"\(producer\)")
                   .and(role().text().eq("actor")))
             .character().text()
             .and(title())))
}

fn q10c() -> String {
    min_row(movie()
        .when(company().country().eq("[us]")
              .and(production_year().gt(1990)))
        .get(cast().when(Cast::note().rx(r"\(producer\)"))
             .character().text()
             .and(title())))
}
